### ISSUE TITLE
HEAD hangs leaving ESTABLISHED connections

### ISSUE THREAD

**Issue by bobrik** [docker/distribution#1182]
This makes registry very unhappy very quick.

GET:

```
$ time curl -v -X GET http://192.168.99.100:5000/v2/myimage/blobs/sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4
*   Trying 192.168.99.100...
* Connected to 192.168.99.100 (192.168.99.100) port 5000 (#0)
> GET /v2/myimage/blobs/sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4 HTTP/1.1
> Host: 192.168.99.100:5000
> User-Agent: curl/7.43.0
> Accept: */*
>
< HTTP/1.1 404 Not Found
< Content-Type: application/json; charset=utf-8
< Docker-Distribution-Api-Version: registry/2.0
< X-Content-Type-Options: nosniff
< Date: Thu, 12 Nov 2015 02:02:43 GMT
< Content-Length: 157
<
{"errors":[{"code":"BLOB_UNKNOWN","message":"blob unknown to registry","detail":"sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4"}]}
* Connection #0 to host 192.168.99.100 left intact

real	0m0.015s
user	0m0.003s
sys	0m0.003s
```

HEAD for the same URI:

```
$ time curl -v -X HEAD http://192.168.99.100:5000/v2/myimage/blobs/sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4
*   Trying 192.168.99.100...
* Connected to 192.168.99.100 (192.168.99.100) port 5000 (#0)
> HEAD /v2/myimage/blobs/sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4 HTTP/1.1
> Host: 192.168.99.100:5000
> User-Agent: curl/7.43.0
> Accept: */*
>
< HTTP/1.1 404 Not Found
< Content-Type: application/json; charset=utf-8
< Docker-Distribution-Api-Version: registry/2.0
< X-Content-Type-Options: nosniff
< Date: Thu, 12 Nov 2015 02:02:56 GMT
< Content-Length: 157
<
^C

real	0m46.279s
user	0m0.004s
sys	0m0.005s
```

`Content-length` looks incorrect here, could be #103 related.

This leads to nginx in front of registry pool to be unable to connect to registry:

```
upstream timed out (110: Connection timed out) while reading response header from upstream, client: 173.245.57.42, server: myregistry, request: "HEAD /v2/scan/blobs/sha256:93b84de2c1ccdd811a768dd608348762c80bdbf6eeec736be0e512fde9f6562b HTTP/1.1", upstream: "http://10.36.17.12:5000/v2/scan/blobs/sha256:93b84de2c1ccdd811a768dd608348762c80bdbf6eeec736be0e512fde9f6562b", host: "myregistry"
```

This makes people see 504 on push.

Seeing this on 2.1.1 and 2.2.0, reproducible locally.

**Comment by stevvooe** [comment]
@bobrik Thanks for the report.

There are a few things going on here.

Let's clear up the non-issues:

1. `Content-Length` is correct. The error message for the 404 response is 157 bytes. This looks correct.
2. `curl`'s implementation of `-X HEAD` is broken. You have to use `-I` or it just leaves the connection open and doesn't send the empty body (I think it terminates with new lines). This problem with curl was actually one of the reasons that caused me to file #103.

To the interesting bits.

I am actually a bit surprised to see that nginx times out further connections. It would be good to see if this could actually recover. If not, we may want to see how many connection slots are being held open to the backend to cause this. A possibility is that the backend is running afoul of the [`fail_timeout`](http://nginx.org/en/docs/http/ngx_http_upstream_module.html#server). Without more information about the configuration, this is just speculation.

That said, there are likely parameters on [`Server`](https://golang.org/pkg/net/http/#Server) for `ReadTimeout` that we have not set, in fear of interrupting long form registry requests. Really what we want is an idle timeout. This [mailing list topic](https://groups.google.com/forum/#!msg/golang-nuts/NX6YzGInRgE/1XmBHwe4DHwJ) touches on the subject.

Would you mind poking around a bit more to identify the root cause? Basically, we need to identify whether the registry isn't shutting inactive connections or if there is some incorrect about the nginx interaction.

**Comment by bobrik** [comment]
Okay, looks like `HEAD` in curl is indeed expected. `ESTABLISHED` connections could be explained by the fact that we use keep-alive connections to upstreams.

I tried calling failing endpoint in a loop and it seems that timeouts (>10s) only happen if you query registry from load balancers. I'll try to get to the root cause tomorrow and update the issue.

Thanks!

**Comment by stevvooe** [comment]
@bobrik Sounds good. Let me know if you need anything.

**Comment by bobrik** [comment]
Okay, so connection timeout of 5s can be triggered from anywhere. Looks like it was correlated to OOM kill of registry container after reaching 512m of RAM. Will try to see if it breaks with higher memory limit.

**Comment by bobrik** [comment]
Okay, looks like unbound memory growth is indeed the issue.

I was able to grow memory up to 1GB (cgroup memory limit) where timeouts happen. I assume that Go runtime tries to allocate more memory, gets errors and tries to run GC more aggressively. Since GC takes over at this point, `accept()` does not finish in time and timeouts happen. OOM follows shortly.

It is fairly easy to replicate by pushing many fat images with lots of layers. We are using Swift backend and registry 2.2.0.

Let me know what kind of information from debug endpoint could be useful. I plan to see where allocations happen.

**Comment by bobrik** [comment]
![image](https://cloud.githubusercontent.com/assets/89186/11156974/bf16fcd6-8a03-11e5-9d62-58ad35292f36.png)

**Comment by stevvooe** [comment]
@bobrik Let's start with the goroutine output during the high memory situation. This will be something like `http://localhost:5001/debug/pprof/goroutine?debug=2`. Grap the heap profile, while you're at it (`http://localhost:5001/debug/pprof/heap?debug=2`).

My guess is there is goroutine leak somewhere.

Do you see this behavior without using the swift driver?

**Comment by bobrik** [comment]
* [goroutine.txt](https://github.com/docker/distribution/files/34437/goroutine.txt)
* [heap.txt](https://github.com/docker/distribution/files/34438/heap.txt)

Haven't tried without swift.

**Comment by stevvooe** [comment]
@bobrik There are large number of goroutines blocked on reads. Could you try it without the swift driver?

**Comment by bobrik** [comment]
I'll try to do that on Monday, let's see what happens.

Also #808 could help us with using s3 instead of swift, though.

**Comment by bobrik** [comment]
Tried running `swift` and `filesystem` registries locally, got the following results after pushing test set of images:

| metric| swift | filesystem |
|---|---|---|
| memory | 103.4mb | 21.65mb |
| goroutines | [goroutine.swift.txt](https://github.com/docker/distribution/files/35353/goroutine.swift.txt) | [goroutine.filesystem.txt](https://github.com/docker/distribution/files/35354/goroutine.filesystem.txt) |
| heap | [heap.swift.txt](https://github.com/docker/distribution/files/35357/heap.swift.txt) | [heap.filesystem.txt](https://github.com/docker/distribution/files/35356/heap.filesystem.txt) |

**Comment by bobrik** [comment]
Now I cannot reproduce growing memory anymore. It stops around 120mb.

I'll create a separate issue if it happens again.

**Comment by stevvooe** [comment]
@bobrik Odd that this went away. We'll keep an eye on this issue in relation to swift driver usage.
